gitlab/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,163 @@
+
+	return resourceGitlabProjectEnvironmentRead(d, meta)
+}
+
+func resourceGitlabProjectEnvironmentRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] read gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return err
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+
+	log.Printf("[DEBUG] Project %s read gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.GetEnvironment(project, environmentIDInt)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			log.Printf("[DEBUG] Project %s gitlab environment %q not found", project, environmentID)
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error getting gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+	d.Set("project", project)
+	d.Set("name", environment.Name)
+	d.Set("state", environment.State)
+
+	return nil
+}
+
+func resourceGitlabProjectEnvironmentUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] update gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return err
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+	if err != nil {
+		return fmt.Errorf("error converting environment ID to int: %v", err)
+	}
+
+	name := d.Get("name").(string)
+	options := gitlab.EditEnvironmentOptions{
+		Name: &name,
+	}
+	if externalURL, ok := d.GetOk("external_url"); ok {
+		options.ExternalURL = gitlab.String(externalURL.(string))
+	}
+
+	log.Printf("[DEBUG] Project %s update gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	if _, _, err := client.Environments.EditEnvironment(project, environmentIDInt, &options); err != nil {
+		return fmt.Errorf("error editing gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	return resourceGitlabProjectEnvironmentRead(d, meta)
+}
+
+func resourceGitlabProjectEnvironmentDelete(d *schema.ResourceData, meta interface{}) error {
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return err
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+	if err != nil {
+		return fmt.Errorf("error converting environment ID to int: %v", err)
+	}
+
+	log.Printf("[DEBUG] Project %s delete gitlab project-level environment %v", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	_, err = client.Environments.StopEnvironment(project, environmentIDInt)
+	if err != nil {
+		return fmt.Errorf("error stopping gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+	_, err = client.Environments.DeleteEnvironment(project, environmentIDInt)
+	if err != nil {
+		return fmt.Errorf("error deleting gitlab project %q environment %q: %v", project, environmentID, err)
+	}
```

Comment:
Is the StopEnvironment API call idempotent? For example, if the DeleteEnvironment call fails and a user retries the delete, will the StopEnvironment API call fail on the second try? (It shouldn't.)

Reply:
Stop:

> It returns 200 if the environment was successfully stopped, and 404 if the environment does not exist. 

https://docs.gitlab.com/ee/api/environments.html#stop-an-environment

Delete:

> It returns 204 if the environment was successfully deleted, and 404 if the environment does not exist.

https://docs.gitlab.com/ee/api/environments.html#delete-an-environment